Submodule subprojects/wf-utils 25ed62f3..98cb1bbb:
diff --git a/subprojects/wf-utils/meson.build b/subprojects/wf-utils/meson.build
index 43b60e8..94c9f84 100644
--- a/subprojects/wf-utils/meson.build
+++ b/subprojects/wf-utils/meson.build
@@ -76,6 +76,7 @@ headers_parser = [
 'wayfire/parser/action_parser.hpp',
 'wayfire/parser/condition_parser.hpp',
 'wayfire/parser/rule_parser.hpp',
+'wayfire/parser/lambda_rule_parser.hpp',
 ]
 
 headers_rule = [
diff --git a/subprojects/wf-utils/wayfire/rule/lambda_rule.cpp b/subprojects/wf-utils/wayfire/rule/lambda_rule.cpp
index 99c558e..0bc6986 100644
--- a/subprojects/wf-utils/wayfire/rule/lambda_rule.cpp
+++ b/subprojects/wf-utils/wayfire/rule/lambda_rule.cpp
@@ -1,6 +1,7 @@
 #include "wayfire/rule/lambda_rule.hpp"
 #include "wayfire/condition/condition.hpp"
 #include <cstddef>
+#include <cstdint>
 #include <functional>
 #include <memory>
 #include <sstream>
